### ISSUE TITLE
Support of CommonJS for @babel/generator, @babel/traverse and other v8 packages

### ISSUE THREAD

**Issue by coderaiser** [babel/babel#15835]
### 💻

- [ ] Would you like to work on this feature?

### What problem are you trying to solve?

Would be great if `@babel/generator` v8 can be not only imported but also has support `commonjs` support, it will make transition smoother and easier.

![image](https://github.com/babel/babel/assets/1573141/ac428ce8-043d-49bb-b4c7-b903a43eddc4)

### Describe the solution you'd like

Export `commonjs` versions of files in a similar to way used in `@babel/parser` v8

![image](https://github.com/babel/babel/assets/1573141/dad7139d-c982-4bc0-a698-a06266982c2c)

### Describe alternatives you've considered

Since **ESLint** (and lots of other packages) doesn't supports **ESM** and even [`async rules`](https://github.com/eslint/eslint/issues/15394), would be great to have both ways to `import` `@babel/generator`, `@babel/traverse` and other packages as **ESM** and as **CommonJS**.

**Comment by liuxingbaoyu** [comment]
https://github.com/prettier/prettier-vscode/pull/3016
https://github.com/microsoft/vscode/issues/130367

I agree with that. In fact even vscode still doesn't support direct use of ESM, prettier was deeply troubled by it before.

**Comment by nicolo-ribaudo** [comment]
Also note that the reason we had a CJS version for Babel parser was prettier, but now that prettier uses ESM we probably do not need it anymore.

**Comment by coderaiser** [comment]
The reason I'm asking is [`eslint-plugin-putout`](https://github.com/coderaiser/putout/tree/master/packages/eslint-plugin-putout) which gives ability to run Babel-based 🐊[**Putout**](https://github.com/coderaiser/putout) code transformer, which should also migrate to **ESM** and most likely it will work slower as sync **CommonJS** **ESLint** plugin that somehow loads **ESM**.

Anyways,

> Babel is popular enough that we are in a position to push the ecosystem to better support ESM.

that's true, so if you decide to move **ESM**-only, I'll try to figure out something to get things working...

**Comment by nicolo-ribaudo** [comment]
Also, note that ESLint is planning a rewrite that would allow it to support asynchronous parsers/plugins: https://github.com/eslint/eslint/discussions/16557

**Comment by liuxingbaoyu** [comment]
> In Babel 7 we had the problem that some of our dependencies were migrating to ESM and thus we were stuck on older versions (or we had to swap them), I don't want to have the same problem in Babel 8.

I agree with that. But I fear there may be worse obstacles downstream. (e.g. Jest was blocked by Node ESM's memory leaks and segfaults for several years)

> Would be great if Babel Team provide some kind of solution or workaround, I see that there is [deasync](https://www.npmjs.com/package/deasync), and even way to use it, which is completely fine, since it has pre-built binaries:

As much as I like `dasync`, it is currently unreliable. The Node team once had a similar PR, but it was rejected.

> Maybe you can suggest me some way to run async code in a sync way?

https://github.com/un-ts/synckit

You can try this. While it may not be very convenient, it is relatively reliable.

**Comment by coderaiser** [comment]
> https://github.com/un-ts/synckit

As I understand i doesn't support passing functions, since it is worker thread...

OK, here is one of possible solutions: pre-built all **CommonJS** in separate package [@putout/babel](https://github.com/putoutjs/babel/), it will simplify supporting of Babel codebase, and makes ability to use Babel 8 inside **ESLint** and other places which cannot migrate to **ESM** now.

**Comment by coderaiser** [comment]
Unfortunately this way not really works with `@babel/core` which is more then `6mb` bundled, so looks like I'm stuck with Babel v7. For how long it will be supported? Will new features arrive to `v7`? Can this AST changes came to `v7` with some env flag, so `v7` was **CommonJS** and `v8` **ESM**.

**Comment by coderaiser** [comment]
Just made [an upgrade](https://github.com/coderaiser/putout/releases/tag/v31.0.0) to **Babel v8**, also for [editor](https://putout.cloudcmd.io/) and [mobile editor](https://putout.vercel.app/) 🎉.

It works like a charm 🥳 !

**Comment by nicolo-ribaudo** [comment]
Awesome, I'm happy you found a solution :)
